### DISCUSSION TITLE
Update Syntax Colors

### DISCUSSION THREAD

**Discussion by bgregs514** [bgregs514/vscode-tm1#11] [General]
Either something went wrong when I merged the tm1helper language files into the project, or they never looked like they did in the project pages screenshot, but the color scheme could probably use some rework.  Right now things are tinted a sort of yellow color, and I would prefer to have some cleaner, bolder colors to distinguish between operators/functions/etc.

**Comment by TM1Fanboy** [comment]
We could go one step further and use different highlighting for useful and rather useless functions like ForceSkipCheck() 🙃. 
Since IBM will not remove those functions, we could give them a different color and could make the decision a community voting. 
What do you guys think?

**Comment by bgregs514** [comment]
@scrumthing Good point!  I'm going to convert this issue to a discussion.  Once we're settled on colors, themes, and paths forward we can pull the finalized proposal back over as an enhancement.  Thanks!

**Comment by ykud** [comment]
My idea would be to remove the custom theme entirely and see if we can get VSCode standard themes to work their magic based on the syntax definition that we maintain (i.e. if we define a syntax rule for keyword or function, they'd be highlighted in any VScode theme that provides a 'function' definition).  I haven't looked at it in much detail yet, so I may be completely wrong here :)

**Reply by TM1Fanboy** [reply]
I think you are right. I played around with syntax highly in vs code a couple of years ago but lost the json in the meantime.

**Reply by bgregs514** [reply]
That's a good point, and actually I think it would be relatively easy to port back to the "default" VS Code themes.  Out of the box, VS Code has predefined [scopes for the syntax highlighting ](https://code.visualstudio.com/api/language-extensions/syntax-highlight-guide#theming).  All we would have to do is modify the .json language files to point back to these scopes.  Side note, the [Scope Inspector](https://code.visualstudio.com/api/language-extensions/syntax-highlight-guide#scope-inspector) tool is super helpful for this type of development.

I'm not an expert on the theming side of things, but my understanding is that languages in VS Code consist of syntax (regex) .json files for grouping and defining what to highlight, theme .json files for defining which color schemes apply to which group in the syntax file, and snippet .json files for the actual code snippets.

What I _think_ would need to happen is basically throw away (or redefine based off of the standard VS Code theme) the CapreDatum theme files, and then update the scope groups in the syntax files.  Redefining the existing theme files to use default VS Code color schemes may be better in the long run because we can use more natural scope naming conventions.  For example, a default VS Code scope may look like:
```
keyword.control
```
whereas we could define it as:
```
keyword.attribute
```
and apply our standard VS Code color scheme to that.  It would be a custom theme, but just using standard VS Code colors.  That would allow us a little more flexibility with naming the color scopes.

Hopefully that was clear, I was struggling to find a good way to type that out.  Anyway, I'm open to ideas - ultimately as long as it looks nice and clean, I'm fine with anything :smile:

**Comment by TM1Fanboy** [comment]
Just a quick note to ensure I will not forget to mention it. Should we include undocumented functions (including proper snippets)? If so they should be highlighted differently to show they are not proper to functions.

**Comment by EddStuart** [comment]
It would be great to introduce a code formatter or to generate a plugin for something like Prettier

Feature request #13

**Reply by TM1Fanboy** [reply]
> It would be great to introduce a code formatter or to generate a plugin for something like Prettier

Love the idea but for that we would have to agree on a default formatting. :-) 
But if we have those rules it should be easier to do proper linting.

**Reply by bgregs514** [reply]
I like this idea a lot.  I created #13 to track this as a feature request, but let's setup a new discussion thread to work out the details of formatting rules, etc. once we pull this into development.

One thing that I think we would need to implement alongside a code formatter would be the ability to edit the formatting rules. 
 Some people like spaces before and after parenthesis, other people (like me) prefer 8 character indents (that drives most people nuts), etc.

I also agree that this is definitely a good path to take us into proper linting.  I think a good place to start would be identifying the areas of code where formatting preference comes into play i.e. parenthesis, indents, capital "IF" vs. lower-case "if', etc.  In all honesty, this may play a bigger role in rules with the amount of variation in line breaks and placement of the "," between elements.

Anyway, great idea.  Let's keep tracking this one.
